xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java

```diff
@@ -496,60 +607,196 @@ private <T extends ResourceUpdate> void handleResourceUpdate(
       // For State of the World services, notify watchers when their watched resource is missing
       // from the ADS update. Note that we can only do this if the resource update is coming from
       // the same xDS server that the ResourceSubscriber is subscribed to.
-      if (subscriber.serverInfo.equals(args.serverInfo)) {
+      if (controlPlaneClient.equals(subscriber.controlPlaneClient)) {
         subscriber.onAbsent(processingTracker);
       }
     }
   }
 
+  private void shutdownLowerPriorityCpcs(ControlPlaneClient activatedCpc) {
+    // For each authority, remove any control plane clients, with lower priority than the activated
+    // one, from activeCpClients storing them all in cpcsToSutdown.
+    List<String> authsForServerInfo = getAuthoritiesForServerInfo(activatedCpc.getServerInfo());
+    Set<ControlPlaneClient> cpcsToSutdown = new HashSet<>();
+    for (String authority : authsForServerInfo) {
+      List<ControlPlaneClient> cpcsForAuth = activeCpClients.get(authority);
+      int index = cpcsForAuth.indexOf(activatedCpc);
+      if (index > -1 && index < cpcsForAuth.size() - 1) {
```

Comment:
nit: Would you mind removing `index < cpcsForAuth.size() - 1` here? It just makes the code harder to read, as it is redundant with the next two lines.

If the point is an optimization, move the add as `cpsToShutdown.add(cpcsForAuth.remove(i))` (or equivalent) into the next `for` loop, which is also more clear.

Reply:
done